kuka_rsi_hw_interface/src/rsi_command_handler.cpp

```diff
@@ -0,0 +1,325 @@
+  xml::XMLElement Override("Override");
+  Override.GetParams()["Override"] = xml::XMLParam(xml::XMLType::LONG);
+  state_data_structure_.GetChilds().emplace_back(Out);
+  state_data_structure_.GetChilds().emplace_back(Override);
+
+  command_data_structure_.GetParams()["Type"] = xml::XMLParam(xml::XMLType::STRING);
+  xml::XMLElement RKorr("RKorr");
+  RKorr.GetParams()["X"] = xml::XMLParam(xml::XMLType::DOUBLE);
+  RKorr.GetParams()["Y"] = xml::XMLParam(xml::XMLType::DOUBLE);
+  RKorr.GetParams()["Z"] = xml::XMLParam(xml::XMLType::DOUBLE);
+  RKorr.GetParams()["A"] = xml::XMLParam(xml::XMLType::DOUBLE);
+  RKorr.GetParams()["B"] = xml::XMLParam(xml::XMLType::DOUBLE);
+  RKorr.GetParams()["C"] = xml::XMLParam(xml::XMLType::DOUBLE);
+  xml::XMLElement DiO("DiO");
+  DiO.GetParams()["DiO"] = xml::XMLParam(xml::XMLType::LONG);
+  command_data_structure_.GetChilds().emplace_back(RKorr);
+  command_data_structure_.GetChilds().emplace_back(DiO);
+}
+
+bool RSICommandHandler::Decode(char * const buffer, const size_t buffer_size)
+{
+  std::cout << "Decode started" << std::endl;
+  auto buffer_it = buffer;
+  try {
+    decodeNode(state_data_structure_, buffer, buffer_it, buffer_size);
+  } catch (const std::exception & e) {
+    std::cout << "### ERROR ###" << std::endl;
+    std::cerr << e.what() << '\n';
+    return false;
+  }
+  std::cout << "Decode finished" << std::endl;
+  return true;
+}
+
+bool RSICommandHandler::Encode(char * & buffer, const size_t buffer_size)
+{
+  std::cout << "Decode started" << std::endl;
+  auto buffer_it = buffer;
+  try {
+    encodeNode(command_data_structure_, buffer, buffer_it, buffer_size);
+  } catch (const std::exception & e) {
+    std::cout << "### ERROR ###" << std::endl;
+    std::cerr << e.what() << '\n';
+    return false;
+  }
+  std::cout << "Decode finished" << std::endl;
+  return true;
+}
+
+void RSICommandHandler::decodeNode(
+  xml::XMLElement & element, char * const buffer, char * & buffer_it,
+  const size_t buffer_size)
+{
+  xml::XMLString node_name(buffer, 0);
+  // fast forward for the first open bracket
+  for (; *buffer_it != '<'; buffer_it++) {
+    if ((int)(buffer_it - buffer) >= (int)buffer_size) {
+      throw std::range_error("Out of the buffers range");
+    }
+  }
+  buffer_it++;
+  // Validate nodes name
+  if (!element.IsNameValid(node_name, buffer_it)) {
+    auto i = std::sprintf(err_buf_, "The detected name (");
+    std::snprintf(err_buf_ + i, node_name.length_ + 1, "%s", node_name.data_ptr_);
+    i += node_name.length_;
+    std::sprintf(
+      err_buf_ + i, ") does not match the elements name %s.",
+      element.GetName().c_str());
+    throw std::logic_error(err_buf_);
+  }
+
+  // Validate nodes params
+  size_t numOfParam = 0;
+  bool isBaseLessNode = false;
+  bool isNoMoreParam = false;
+  xml::XMLString node_param;
+  while (!isNoMoreParam && (size_t)(buffer_it - buffer) < buffer_size) {
+    // fast forward to the next non-space character
```

Comment:
It is ok for now, but I think this cannot handle more special whitespace characters, like tab, line break etc

Reply:
later, add to issue